pkg/vmssrollout/upgrade.go

```diff
@@ -0,0 +1,325 @@
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+)
+
+type ImageReference struct {
+	ID      string
+	SKU     string
+	Version string
+}
+
+type PluginInterface interface {
+	WaitForReadiness(ctx context.Context, nodeName string) error
+}
+
+type VMSSUpgrader struct {
+	SubscriptionID string
+	ResourceGroup  string
+	// Name of the VMSS to upgrade.
+	Name string
+
+	// provides an interface for configuring
+	// waiting for the readiness of a VM.
+	Plugin PluginInterface
+
+	// upgrade parameters
+	DesiredReplicas int64
+	ImageRef        *ImageReference
+	Drain           bool
+	InPlace         bool
+	MaxSurge        int64
+	MaxUnavailable  int64
+
+	ssc compute.VirtualMachineScaleSetsClient
+	vmc compute.VirtualMachineScaleSetVMsClient
+}
+
+func (u *VMSSUpgrader) clientSetup() error {
+	authorizer, err := auth.NewAuthorizerFromEnvironment()
+	if err != nil {
+		return err
+	}
+
+	ssc := compute.NewVirtualMachineScaleSetsClient(u.SubscriptionID)
+	ssc.Authorizer = authorizer
+	u.ssc = ssc
+
+	vmc := compute.NewVirtualMachineScaleSetVMsClient(u.SubscriptionID)
+	vmc.Authorizer = authorizer
+	u.vmc = vmc
+	return nil
+}
+
+func (u *VMSSUpgrader) Upgrade() error {
+	if err := u.clientSetup(); err != nil {
+		return err
+	}
+
+	vmss, err := u.ssc.Get(context.Background(), u.ResourceGroup, u.Name)
+	if err != nil {
+		return fmt.Errorf("cannot get VMSS %q in resource group %q: %v", u.Name, u.ResourceGroup, err)
+	}
+
+	if vmss, err = u.updateImageVMSS(vmss); err != nil {
+		return err
+	}
+
+	log.Printf("VMSS %q current capacity at %d", u.Name, *vmss.Sku.Capacity)
+
+	vmsToUpgrade, err := u.upgradable()
+	if err != nil {
+		return fmt.Errorf("cannot list upgradable VMs in VMSS %q: %v", u.Name, err)
+	}
+
+	for _, vmToUpgrade := range vmsToUpgrade {
+		// if possible, scale up first
+		canScaleUp, err := u.canScaleUp()
+		if err != nil {
+			return err
+		}
+		if canScaleUp {
+			log.Printf("Scale up VMSS %q", u.Name)
+			*vmss.Sku.Capacity = u.DesiredReplicas + u.MaxSurge
+			res, err := u.ssc.CreateOrUpdate(context.Background(), u.ResourceGroup, u.Name, vmss)
+			if err != nil {
+				return fmt.Errorf("cannot scale up VMSS %q: %v", u.Name, err)
+			}
+			if err := res.WaitForCompletion(context.Background(), u.ssc.BaseClient.Client); err != nil {
+				return fmt.Errorf("cannot wait for VMSS %q scale up: %v", u.Name, err)
+			}
+			if vmss, err = res.Result(u.ssc); err != nil {
+				return fmt.Errorf("error during VMSS %q scale up: %v", u.Name, err)
+			}
+			// Wait for VM readiness; this may vary between VMSSs, eg. a master VM
+			// is ready when the API server pod running on it is ready whereas a compute
+			// VM is ready when it has joined the cluster as a Ready k8s node.
+			// TODO: Figure out the new node name
+			//if err := u.Plugin.WaitForReadiness(context.Background(), nodeName); err != nil {
+			//	return err
+			//}
+		}
+
+		canScaleDown, err := u.canScaleDown()
```

Comment:
why we are checking if we can scale down? Is there any reason why we should not be able to scale down?